projects/hathor-wallet-headless/0001-multisig-create-mint-melt-token.md

```diff
@@ -0,0 +1,327 @@
+}
+```
+
+The purpose of the command is only produce the `txHex` not signed and nothing more, it will not push the transaction to the network. The returned `txHex` will be handled by the coordinator service, and later on by the federation service, which is responsible to sign the transaction.
+
+With this specification we cover the possibility to create custom tokens with or without mint and melt authorities.
+
+## Mint a custom token
+
+A single wallet can [mint a token](https://hathor.gitbook.io/hathor/guides/headless-wallet/creating-tokens/mint-tokens) using `POST:/wallet/mint-token` endpoint. This operation creates another mint authority by default.
+
+For the MultiSig wallet the endpoint keeps the same command:
+`POST:/wallet/p2sh/tx-proposal/mint-token`
+
+The data fields are:
+* `token`*
+* `amount`*
+* `address`*
+* `change_address`
+* `mint_authority_address`
+* `allow_external_mint_authority_address`
+
+*`*` as required field 
+
+The response scheme for success:
+```ts
+{
+  "success": boolean,
+  "txHex": string
+}
+```
+
+The purpose of the command is only produce the `txHex` not signed and nothing more, it will not push the transaction to the network. The returned `txHex` will be handled by the coordinator service, and later on by the federation service, which is responsible to sign the transaction.
+
+With this specification we cover the possibility to mint tokens, send it to another wallet, create or not another mint authority, and send the authority to another wallet.
+
+## Melt a custom token
+
+A single wallet can melt a token using `POST:/wallet/melt-token` endpoint. This operation creates another melt authority by default.
+
+For the MultiSig wallet the endpoint keeps the same command:
+`POST:/wallet/p2sh/tx-proposal/melt-token`
+
+The data fields are:
+* `token`*
+* `amount`*
+* `address`*
+* `change_address`
+* `melt_authority_address`
+* `allow_external_melt_authority_address`
+
+*`*` as required field 
+
+The response scheme for success:
+```ts
+{
+  "success": boolean,
+  "txHex": string
+}
+```
+
+The purpose of the command is only produce the `txHex` not signed and nothing more, it will not push the transaction to the network. The returned `txHex` will be handled by the coordinator service, and later on by the federation service, which is responsible to sign the transaction.
+
+With this specification we cover the possibility to melt tokens, send it to another wallet, create or not another melt authority, and send the authority to another wallet.
+
+## Inspect transaction with wallet metadata
+
+There is an endpoint to [inspect the transaction](https://hathor.gitbook.io/hathor/guides/multisig-wallet/sending-transactions#5.-check-the-multisig-transaction) by decoding the transaction hexadecimal encode in `POST:/wallet/decode`. However, it doesn't bring signature information, nor wallet information with it.
+
+For the MultiSig wallet we have the `inspect` command:
+`POST:/wallet/p2sh/tx-proposal/inspect`
```

Comment:
Is there any reason to create a new endpoint, instead of just using the same (`/wallet/decode`) expanding it to return a more detailed data? Maybe even with a parameter `include_metadata` if we feel the common use is to get only the simple data.

Reply:
I don't think the syntax consistency is broken when we use the `/wallet/decode` api, we will improve an api made to work with both p2pkh and p2sh wallets.

Reply:
The reason for my comment is that the tx decode doesn't change if we are using a multisig wallet or not. We have different endpoints for sending transactions because there are some specific steps when sending a tx with a multisig wallet but decoding a transaction is always the same.

Reply:
Well, if the API consistency is not a thing in the way I have though, and we are ok to change the scheme for the response in the presence of "include_metadata" param in the request, we can extend the `/wallet/decode`.

Particularly I consider it a breaking change, and it would be difficult to describe and explain this subtle change in the scheme with the introduction of a request param.

Reply:
My proposition to solve this topic is:
- extend the `/wallet/decode` to multisig -- if needed
- create a second endpoint `/wallet/enriched-decode` that contains wallet metadata

For future should we think to have a GraphQL like endpoint?

Reply:
I like having a second endpoint for this, but I think `/wallet/inspect` seems more to the point. What do you think?

Reply:
The information added as metadata seems to be useful even for normal wallets and use. I prefer that we expand our response object to add this new data. With that we don't create a new endpoint specifically for metadata (this would require users to send two requests to get full data), and we don't add breaking changes, because the new data will be a complement of the current response data.

Reply:
You are not considering this addition as a breaking change because nothing would change for the calls without the `include_metadata` param, right?

I consider any change in the response scheme a breaking change, even complementing ones, because one may validate the response scheme before use it.

I don't see any problem for us to proceed with the `include_metadata` solution. But we should agree that any change in the `include_metadata` scheme in the future would be a breaking change, except if we add another mechanism to not disrupt the current scheme usage, and to avoid these exceptional mechanisms I believe we need a versioning system for the API.

But coming back to the problem of documentation to include two schemes for the same endpoint, the OpenAPI v3 supports the ["Discriminator Object"](https://swagger.io/specification/#discriminator-object), which is the specification for multiple schemes, however it doesn't support deliver an specific scheme related to a param and we should write a very clear documentation about this scheme change attached to the `include_metadata`.

---

Facing these limitations of OpenAPI, the extra care with the documentation to bring clarity and the creation of exceptional mechanism to keep backward compatibility, I'm starting to gain resistance toward this solution.

The creation of `/wallet/inspect` endpoint seen now pretty neat, more simple. We can still have only one function to decode.

Reply:
I have [updated the design](https://github.com/HathorNetwork/rfcs/blob/24ec11b13901870c115774bb117ba962f8079851/projects/hathor-wallet-headless/0001-multisig-create-mint-melt-token.md#decode-transaction-with-wallet-metadata) with the decision we made together.